File: src/utils/phone-validation.ts
/**
 * Phone Validation Utilities for South African Phone Numbers
 *
 * Supports two formats:
 * - E.164 International: +27821234567
 * - South African Local: 0821234567
 */

/**
 * Detects if input string is a phone number
 * Supports: E.164 format (+27...) and SA format (0...)
 */
export function isPhoneNumber(input: string): boolean {
  const trimmed = input.trim()

  // E.164 format: starts with +, followed by 10-15 digits
  const e164Regex = /^\+\d{10,15}$/

  // South African format: starts with 0, followed by 9 digits
  const saRegex = /^0\d{9}$/

  return e164Regex.test(trimmed) || saRegex.test(trimmed)
}

/**
 * Validates phone number format
 * Returns true if the phone number is valid for South Africa
 */
export function isValidPhoneNumber(phone: string): boolean {
  const trimmed = phone.trim()

  // Must be E.164 or SA format
  if (!isPhoneNumber(trimmed)) return false

  // Additional validation: SA numbers must have valid prefixes
  if (trimmed.startsWith('0')) {
    // Valid SA mobile prefixes: 06x, 07x, 08x
    const prefix = trimmed.substring(0, 3)
    return /^0[678]\d/.test(prefix)
  }

  // E.164: check it starts with +27 and has correct length (12 chars total)
  if (trimmed.startsWith('+27')) {
    return trimmed.length === 12
  }

  // Other international numbers - basic validation
  return trimmed.startsWith('+') && trimmed.length >= 11 && trimmed.length <= 16
}

/**
 * Formats phone number for display
 * Examples:
 *   "+27821234567" → "+27 82 123 4567"
 *   "0821234567" → "082 123 4567"
 */
export function formatPhoneNumber(phone: string): string {
  const trimmed = phone.trim()

  if (trimmed.startsWith('+27')) {
    // +27 82 123 4567
    return trimmed.replace(/(\+27)(\d{2})(\d{3})(\d{4})/, '$1 $2 $3 $4')
  }

  if (trimmed.startsWith('0')) {
    // 082 123 4567
    return trimmed.replace(/(\d{3})(\d{3})(\d{4})/, '$1 $2 $3')
  }

  // Return as-is if format not recognized
  return phone
}

/**
 * Masks phone number for security/privacy
 * Example: "+27821234567" → "+27 82 *** 4567"
 */
export function maskPhoneNumber(phone: string): string {
  const trimmed = phone.trim()

  if (trimmed.startsWith('+27')) {
    // +27 82 *** 4567
    return trimmed.replace(/(\+27)(\d{2})\d{3}(\d{4})/, '$1 $2 *** $3')
  }

  if (trimmed.startsWith('0')) {
    // 082 *** 4567
    return trimmed.replace(/(\d{3})\d{3}(\d{4})/, '$1 *** $2')
  }

  // Return partially masked for unknown formats
  const len = phone.length
  if (len > 4) {
    return phone.substring(0, 3) + '***' + phone.substring(len - 4)
  }

  return phone
}

/**
 * Normalizes phone to E.164 format for API calls
 * "0821234567" → "+27821234567"
 */
export function normalizePhoneNumber(phone: string): string {
  const trimmed = phone.trim()

  // Already E.164 format
  if (trimmed.startsWith('+')) return trimmed

  // Convert SA format to E.164
  if (trimmed.startsWith('0')) {
    return '+27' + trimmed.substring(1)
  }

  // Return as-is if format not recognized
  return phone
}

File: app/utils/phone_validation.py
import re
from typing import Optional, Literal, Tuple


def detect_login_identifier_type(identifier: str) -> Tuple[Literal["email", "phone", "unknown"], Optional[str]]:
    """
    Detect if the login identifier is an email or phone number.

    Args:
        identifier: User input (email or phone)

    Returns:
        Tuple of (identifier_type, normalized_value):
        - ("email", "user@example.com") for valid email
        - ("phone", "+27821234567") for valid phone (returns normalized E.164 format)
        - ("unknown", None) if neither valid email nor phone
    """
    if not identifier or not identifier.strip():
        return ("unknown", None)

    identifier = identifier.strip()

    # Check if it's an email (basic pattern matching)
    # Pattern: something@something.domain
    email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if re.match(email_pattern, identifier):
        return ("email", identifier)

    # Check if it could be a phone number
    # Remove common formatting characters
    cleaned = re.sub(r'[\s\-\(\)\.]', '', identifier)

    # Check for phone patterns:
    # - Starts with + (international)
    # - Starts with 0 (South African local)
    # - Just digits (might be phone without formatting)
    if cleaned.startswith('+') or cleaned.startswith('0') or cleaned.isdigit():
        # Validate and normalize
        try:
            normalized = normalize_phone_number(identifier)
            return ("phone", normalized)
        except ValueError:
            pass

    return ("unknown", None)


def normalize_phone_number(phone: str) -> str:
    """
    Normalize phone number to E.164 format for AWS Cognito.

    Args:
        phone: Phone number in various formats

    Returns:
        Phone number in E.164 format (+country_code followed by number)

    Raises:
        ValueError: If phone number format is invalid
    """
    if not phone:
        raise ValueError("Phone number is required")

    # Remove all spaces, dashes, parentheses, and other formatting
    cleaned = re.sub(r'[\s\-\(\)\.]', '', phone.strip())

    # If already in E.164 format (starts with +), validate and return
    if cleaned.startswith('+'):
        if not re.match(r'^\+[1-9]\d{7,14}$', cleaned):  # 8-15 digits total (country code + number)
            raise ValueError("Invalid international phone number format. Use format: +27123456789")
        return cleaned

    # Handle South African numbers (most common for this platform)
    if cleaned.startswith('0'):
        # Convert local SA format (0XX) to international (+27XX)
        if len(cleaned) == 10:  # Standard SA mobile: 0123456789
            sa_number = '+27' + cleaned[1:]
            if not re.match(r'^\+27[1-9]\d{8}$', sa_number):
                raise ValueError("Invalid South African phone number format")
            return sa_number
        else:
            raise ValueError("South African phone numbers should be 10 digits starting with 0")

    # Handle numbers that might be missing country code
    if cleaned.isdigit() and len(cleaned) == 9:
        # Assume South African number without leading 0
        sa_number = '+27' + cleaned
        if not re.match(r'^\+27[1-9]\d{8}$', sa_number):
            raise ValueError("Invalid phone number format")
        return sa_number

    # If none of the above patterns match
    raise ValueError(
        "Invalid phone number format. Please use one of these formats:\n"
        "- International: +27123456789\n"
        "- South African: 0123456789"
    )


def validate_phone_number(phone: str) -> bool:
    """
    Validate if phone number can be normalized to E.164 format.

    Args:
        phone: Phone number to validate

    Returns:
        True if valid, False otherwise
    """
    try:
        normalize_phone_number(phone)
        return True
    except ValueError:
        return False


def get_phone_validation_error(phone: str) -> Optional[str]:
    """
    Get specific validation error message for phone number.

    Args:
        phone: Phone number to validate

    Returns:
        Error message if invalid, None if valid
    """
    try:
        normalize_phone_number(phone)
        return None
    except ValueError as e:
        return str(e)